Tests/SwiftGRPCNIOTests/GRPCChannelHandlerTests.swift

```diff
@@ -0,0 +1,196 @@
+  }
+
+  func testInternalErrorStatusIsReturnedWhenSendingTrailersInRequest() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 2) { channel in
+      // We have to use "Collect" (client streaming) as the tests rely on `EmbeddedChannel` which runs in this thread.
+      // In the current server implementation, responses from unary calls send a status immediately after sending the response.
+      // As such, a unary "Get" would return an "ok" status before the trailers would be sent.
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Collect")
+      try channel.writeInbound(HTTPServerRequestPart.head(requestHead))
+      try channel.writeInbound(HTTPServerRequestPart.body(gRPCMessage(channel: channel)))
+
+      var trailers = HTTPHeaders()
+      trailers.add(name: "foo", value: "bar")
+      try channel.writeInbound(HTTPServerRequestPart.end(trailers))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractStatus(responses[1])) { status in
+      XCTAssertEqual(status.code, .internalError)
+    }
+  }
+
+  func testOnlyOneStatusIsReturned() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 3) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Get")
+      try channel.writeInbound(HTTPServerRequestPart.head(requestHead))
+      try channel.writeInbound(HTTPServerRequestPart.body(gRPCMessage(channel: channel)))
+
+      // Sending trailers with `.end` should trigger an error. However, writing a message to a unary call
+      // will trigger a response and status to be sent back. Since we're using `EmbeddedChannel` this will
+      // be done before the trailers are sent. If a 4th resposne were to be sent (for the error status) then
+      // the test would fail.
+
+      var trailers = HTTPHeaders()
+      trailers.add(name: "foo", value: "bar")
+      try channel.writeInbound(HTTPServerRequestPart.end(trailers))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractMessage(responses[1]))
+    XCTAssertNoThrow(try extractStatus(responses[2])) { status in
+      XCTAssertEqual(status.code, .ok)
+    }
+  }
+
+  override func waitForGRPCChannelHandlerResponses(
+    count: Int,
+    servicesByName: [String: CallHandlerProvider] = GRPCChannelHandlerResponseCapturingTestCase.echoProvider,
+    callback: @escaping (EmbeddedChannel) throws -> Void
+  ) throws -> [RawGRPCServerResponsePart] {
+    return try super.waitForGRPCChannelHandlerResponses(count: count, servicesByName: servicesByName) { channel in
+      _ = channel.pipeline.addHandlers(HTTP1ToRawGRPCServerCodec(), first: true)
+        .thenThrowing { _ in try callback(channel) }
+    }
+  }
+}
+
+// Assert the given expression does not throw, and validate the return value from that expression.
+public func XCTAssertNoThrow<T>(
```

Comment:
Terrific idea! I've always been bothered that `XCTAssertNoThrow` could not give you the result of its call. How about moving this to a `TestHelpers` file (or to `GRPCChannelHandlerResponseCapturingTestCase`), though?